pkg/operator/ceph/object/objectstore.go

```diff
@@ -88,18 +98,89 @@ func deleteRealmAndPools(context *Context, spec cephv1.ObjectStoreSpec) error {
 	return nil
 }
 
+func checkZoneIsMaster(context *Context, realmArg string, zoneGroupArg string, zoneArg string) (bool, error) {
+	zoneGroupOutput, err := RunAdminCommandNoRealm(context, "zonegroup", "get", realmArg, zoneGroupArg)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get rgw zone group")
+	}
+	masterZoneID, err := decodeMasterZoneID(zoneGroupOutput)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse master zone id")
+	}
+
+	zoneOutput, err := RunAdminCommandNoRealm(context, "zone", "get", realmArg, zoneGroupArg, zoneArg)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get rgw zone")
+	}
+	zoneID, err := decodeID(zoneOutput)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse zone id")
+	}
+
+	if zoneID == masterZoneID {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func getZoneEndpoints(context *Context, realmArg string, zoneGroupArg string, zoneName string, serviceEndpoint string) (string, error) {
+	zoneGroupOutput, err := RunAdminCommandNoRealm(context, "zonegroup", "get", realmArg, zoneGroupArg)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get rgw zone group")
+	}
+	zones, err := decodeZonesList(zoneGroupOutput)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse zones list")
+	}
+
+	numZones := len(zones)
+	zoneEndpoints := ""
+	for i := 0; i < numZones; i++ {
+		if zones[i].Name == zoneName {
+			numEndpoints := len(zones[i].Endpoints)
+			for j := 0; j < numEndpoints; j++ {
+				// in case object-store operator code is rereconciled, zone modify could get run again with serviceEndpoint added again
+				if zones[i].Endpoints[j] != serviceEndpoint {
+					zoneEndpoints += zones[i].Endpoints[j] + ","
+				}
+			}
+			break
+		}
+	}
+
+	zoneEndpoints += serviceEndpoint
```

Comment:
Could the serviceEndpoint ever be the same as one of the zoneEndpoints already appended? Or what is the difference?

Reply:
zoneEndpoints is the list of the endpoints already in the Zone, not including the serviceEndpoint. serviceEndpoint is the endpoint of the object-store service being added to the Zone. There is a check in the nested loop above that makes sure not to include serviceEndpoint into zoneEndpoints when it is being modified in the nested loop. 

Before I added that check, there was one instance in my local testing the objectstore reconciled again. On this second reconciliation the zone modify with the serviceEndpoint had already been run and the serviceEndpoint was already in the zoneEndpoints because the `zone modify` command had already been run in the previous reconciliation. this resulted in the serviceEndpoint being added twice. It's my understanding that every reconciliation should be idempotent.

Reply:
Agreed, idempotent is definitely a requirement, thanks for the explanation.